### PR TITLE
[REVIEW] Support truncated fractions in cudf::strings::to_timestamp

### LINKED ISSUE #6219: [BUG] to_timestamps not handling format correctly when there is a truncated string
**Describe the bug**
We use `to_timestamps` to convert a string into a datetime column. This is primarily done incase of type-casting a string column to datetime column. In this case we try to infer format from first row and pass the format to `to_timestamps` api. It appears to be that when `%.9f` exists the corresponding value is not being inferred correctly.

**Steps/Code to reproduce bug**
```python
>>> import cudf
>>> s = cudf.Series([ "1970-01-01 00:00:00.000001", "1970-01-01 00:00:00.000001000"])
>>> s
0       1970-01-01 00:00:00.000001
1    1970-01-01 00:00:00.000001000
dtype: object

# CORRECT CASE
>>> s.astype('datetime64[ns]')
# Inferred Format from 1st Value is: %Y-%m-%d %H:%M:%S.%f
0   1970-01-01 00:00:00.000001
1   1970-01-01 00:00:00.000001
dtype: datetime64[ns]
>>> s.to_pandas().astype('datetime64[ns]')
0   1970-01-01 00:00:00.000001
1   1970-01-01 00:00:00.000001
dtype: datetime64[ns]

# INCORRECT CASE - When the inferred format has %9f
>>> s = cudf.Series(["1970-01-01 00:00:00.000001000", "1970-01-01 00:00:00.000001"])
>>> s
0    1970-01-01 00:00:00.000001000
1       1970-01-01 00:00:00.000001
dtype: object
>>> s.astype('datetime64[ns]')
# Inferred Format from 1st Value is: %Y-%m-%d %H:%M:%S.%9f
0   1970-01-01 00:00:00.000001000
1   1970-01-01 00:00:00.000000001 # <- Incorrect inference
dtype: datetime64[ns]
>>> s.to_pandas().astype('datetime64[ns]')
0   1970-01-01 00:00:00.000001
1   1970-01-01 00:00:00.000001
dtype: datetime64[ns]

```
**Expected behavior**
The expected behavior is the result should be:
```python
>>> s = cudf.Series(["1970-01-01 00:00:00.000001000", "1970-01-01 00:00:00.000001"])
>>> s
0    1970-01-01 00:00:00.000001000
1       1970-01-01 00:00:00.000001
dtype: object
>>> s.astype('datetime64[ns]')
0   1970-01-01 00:00:00.000001000
1   1970-01-01 00:00:00.000001000
dtype: datetime64[ns]
```

**Environment overview (please complete the following information)**
 - Environment location: Bare-metal
 - Method of cuDF install: from source

**Environment details**
<details><summary>Click here to see environment details</summary><pre>
     
     **git***
     commit b3c642f6e5fa15d5f422d82ad1c79acd7bee436d (HEAD -> 1892, origin/1892)
     Author: galipremsagar <sagarprem75@gmail.com>
     Date:   Sun Sep 13 03:50:17 2020 -0700
     
     adapt changed signatures
     **git submodules***
     
     ***OS Information***
     DISTRIB_ID=Ubuntu
     DISTRIB_RELEASE=18.04
     DISTRIB_CODENAME=bionic
     DISTRIB_DESCRIPTION="Ubuntu 18.04.4 LTS"
     NAME="Ubuntu"
     VERSION="18.04.4 LTS (Bionic Beaver)"
     ID=ubuntu
     ID_LIKE=debian
     PRETTY_NAME="Ubuntu 18.04.4 LTS"
     VERSION_ID="18.04"
     HOME_URL="https://www.ubuntu.com/"
     SUPPORT_URL="https://help.ubuntu.com/"
     BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
     PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
     VERSION_CODENAME=bionic
     UBUNTU_CODENAME=bionic
     Linux dt07 4.15.0-76-generic #86-Ubuntu SMP Fri Jan 17 17:24:28 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
     
     ***GPU Information***
     Sun Sep 13 10:57:00 2020
     +-----------------------------------------------------------------------------+
     | NVIDIA-SMI 440.64.00    Driver Version: 440.64.00    CUDA Version: 10.2     |
     |-------------------------------+----------------------+----------------------+
     | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
     | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
     |===============================+======================+======================|
     |   0  Tesla T4            On   | 00000000:3B:00.0 Off |                    0 |
     | N/A   48C    P8    11W /  70W |      0MiB / 15109MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   1  Tesla T4            On   | 00000000:5E:00.0 Off |                    0 |
     | N/A   37C    P8     9W /  70W |      0MiB / 15109MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   2  Tesla T4            On   | 00000000:AF:00.0 Off |                    0 |
     | N/A   32C    P8    10W /  70W |      0MiB / 15109MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     |   3  Tesla T4            On   | 00000000:D8:00.0 Off |                    0 |
     | N/A   32C    P8    11W /  70W |      0MiB / 15109MiB |      0%      Default |
     +-------------------------------+----------------------+----------------------+
     
     +-----------------------------------------------------------------------------+
     | Processes:                                                       GPU Memory |
     |  GPU       PID   Type   Process name                             Usage      |
     |=============================================================================|
     |  No running processes found                                                 |
     +-----------------------------------------------------------------------------+
     
     ***CPU***
     Architecture:        x86_64
     CPU op-mode(s):      32-bit, 64-bit
     Byte Order:          Little Endian
     CPU(s):              64
     On-line CPU(s) list: 0-63
     Thread(s) per core:  2
     Core(s) per socket:  16
     Socket(s):           2
     NUMA node(s):        2
     Vendor ID:           GenuineIntel
     CPU family:          6
     Model:               85
     Model name:          Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz
     Stepping:            4
     CPU MHz:             1133.790
     BogoMIPS:            4200.00
     Virtualization:      VT-x
     L1d cache:           32K
     L1i cache:           32K
     L2 cache:            1024K
     L3 cache:            22528K
     NUMA node0 CPU(s):   0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62
     NUMA node1 CPU(s):   1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63
     Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti intel_ppin ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke md_clear flush_l1d
     
     ***CMake***
     /nvme/0/pgali/envs/cudfdev/bin/cmake
     cmake version 3.18.2
     
     CMake suite maintained and supported by Kitware (kitware.com/cmake).
     
     ***g++***
     /usr/bin/g++
     g++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
     Copyright (C) 2017 Free Software Foundation, Inc.
     This is free software; see the source for copying conditions.  There is NO
     warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
     
     
     ***nvcc***
     /usr/local/cuda/bin/nvcc
     nvcc: NVIDIA (R) Cuda compiler driver
     Copyright (c) 2005-2019 NVIDIA Corporation
     Built on Wed_Oct_23_19:24:38_PDT_2019
     Cuda compilation tools, release 10.2, V10.2.89
     
     ***Python***
     /nvme/0/pgali/envs/cudfdev/bin/python
     Python 3.7.8
     
     ***Environment Variables***
     PATH                            : /usr/share/swift/usr/bin:/home/nfs/pgali/bin:/home/nfs/pgali/.local/bin:/nvme/0/pgali/envs/cudfdev/bin:/home/nfs/pgali/anaconda3/bin:/home/nfs/pgali/anaconda3/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/lib/jvm/default-java/bin:/usr/share/sbt-launcher-packaging/bin/sbt-launch.jar/bin:/usr/lib/spark/bin:/usr/lib/spark/sbin:/usr/local/cuda/bin
     LD_LIBRARY_PATH                 : /usr/local/cuda/lib64::/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64
     NUMBAPRO_NVVM                   :
     NUMBAPRO_LIBDEVICE              :
     CONDA_PREFIX                    : /nvme/0/pgali/envs/cudfdev
     PYTHON_PATH                     :
     
     ***conda packages***
     /home/nfs/pgali/anaconda3/bin/conda
     # packages in environment at /nvme/0/pgali/envs/cudfdev:
     #
     # Name                    Version                   Build  Channel
     _libgcc_mutex             0.1                 conda_forge    conda-forge
     _openmp_mutex             4.5                       1_gnu    conda-forge
     abseil-cpp                20200225.2           he1b5a44_2    conda-forge
     alabaster                 0.7.12                     py_0    conda-forge
     appdirs                   1.4.3                      py_1    conda-forge
     argon2-cffi               20.1.0           py37h8f50634_1    conda-forge
     arrow-cpp                 1.0.1           py37h1234567_1_cuda    conda-forge
     arrow-cpp-proc            1.0.1                      cuda    conda-forge
     attrs                     20.2.0             pyh9f0ad1d_0    conda-forge
     aws-sdk-cpp               1.7.164              hba45d7a_2    conda-forge
     babel                     2.8.0                      py_0    conda-forge
     backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
     backports                 1.0                        py_2    conda-forge
     backports.functools_lru_cache 1.6.1                      py_0    conda-forge
     black                     19.10b0                    py_4    conda-forge
     bleach                    3.1.5              pyh9f0ad1d_0    conda-forge
     bokeh                     2.2.1            py37hc8dfbb8_0    conda-forge
     boost-cpp                 1.72.0               h7b93d67_2    conda-forge
     brotli                    1.0.9                he1b5a44_0    conda-forge
     brotlipy                  0.7.0           py37h8f50634_1000    conda-forge
     bzip2                     1.0.8                h516909a_3    conda-forge
     c-ares                    1.16.1               h516909a_3    conda-forge
     ca-certificates           2020.6.20            hecda079_0    conda-forge
     certifi                   2020.6.20        py37hc8dfbb8_0    conda-forge
     cffi                      1.14.1           py37h2b28604_0    conda-forge
     cfgv                      3.2.0                      py_0    conda-forge
     chardet                   3.0.4           py37hc8dfbb8_1006    conda-forge
     clang                     8.0.1                hc9558a2_2    conda-forge
     clang-tools               8.0.1                hc9558a2_2    conda-forge
     clangxx                   8.0.1                         2    conda-forge
     click                     7.1.2              pyh9f0ad1d_0    conda-forge
     cloudpickle               1.6.0                      py_0    conda-forge
     cmake                     3.18.2               h5c55442_0    conda-forge
     cmake_setuptools          0.1.3                      py_0    rapidsai
     commonmark                0.9.1                      py_0    conda-forge
     cryptography              3.1              py37hb09aad4_0    conda-forge
     cudatoolkit               10.2.89              h6bb024c_0    nvidia
     cudf                      0.16.0a0+1725.gb3c642f6e.dirty          pypi_0    pypi
     cudnn                     7.6.5                cuda10.2_0
     cupy                      7.8.0            py37h940342b_0    conda-forge
     curl                      7.71.1               he644dc0_5    conda-forge
     cython                    0.29.21          py37h3340039_0    conda-forge
     cytoolz                   0.10.1           py37h516909a_0    conda-forge
     dask                      2.25.0+12.g22224bb0          pypi_0    pypi
     dask-cudf                 0.16.0a0+1704.g03c4d6b8c.dirty          pypi_0    pypi
     decorator                 4.4.2                      py_0    conda-forge
     defusedxml                0.6.0                      py_0    conda-forge
     distlib                   0.3.1              pyh9f0ad1d_0    conda-forge
     distributed               2.25.0+7.ga8032534          pypi_0    pypi
     dlpack                    0.3                  he1b5a44_1    conda-forge
     docutils                  0.16             py37hc8dfbb8_1    conda-forge
     double-conversion         3.1.5                he1b5a44_2    conda-forge
     editdistance              0.5.3            py37h3340039_1    conda-forge
     entrypoints               0.3             py37hc8dfbb8_1001    conda-forge
     expat                     2.2.9                he1b5a44_2    conda-forge
     fastavro                  1.0.0.post1      py37h8f50634_0    conda-forge
     fastrlock                 0.5              py37h3340039_0    conda-forge
     filelock                  3.0.12             pyh9f0ad1d_0    conda-forge
     flake8                    3.8.3                      py_1    conda-forge
     flatbuffers               1.12.0               he1b5a44_0    conda-forge
     freetype                  2.10.2               he06d7ca_0    conda-forge
     fsspec                    0.8.0                      py_0    conda-forge
     future                    0.18.2           py37hc8dfbb8_1    conda-forge
     gflags                    2.2.2             he1b5a44_1004    conda-forge
     glog                      0.4.0                h49b9bf7_3    conda-forge
     gmp                       6.2.0                he1b5a44_2    conda-forge
     grpc-cpp                  1.30.2               heedbac9_0    conda-forge
     heapdict                  1.0.1                      py_0    conda-forge
     hypothesis                5.28.0                     py_0    conda-forge
     icu                       67.1                 he1b5a44_0    conda-forge
     identify                  1.5.0              pyh9f0ad1d_0    conda-forge
     idna                      2.10               pyh9f0ad1d_0    conda-forge
     imagesize                 1.2.0                      py_0    conda-forge
     importlib-metadata        1.7.0            py37hc8dfbb8_0    conda-forge
     importlib_metadata        1.7.0                         0    conda-forge
     iniconfig                 1.0.1              pyh9f0ad1d_0    conda-forge
     ipykernel                 5.3.4            py37h43977f1_0    conda-forge
     ipython                   7.18.1           py37hc6149b9_0    conda-forge
     ipython_genutils          0.2.0                      py_1    conda-forge
     isort                     5.0.7            py37hc8dfbb8_0    conda-forge
     jedi                      0.17.2           py37hc8dfbb8_0    conda-forge
     jinja2                    2.11.2             pyh9f0ad1d_0    conda-forge
     jpeg                      9d                   h516909a_0    conda-forge
     jsonschema                3.2.0            py37hc8dfbb8_1    conda-forge
     jupyter_client            6.1.7                      py_0    conda-forge
     jupyter_core              4.6.3            py37hc8dfbb8_1    conda-forge
     krb5                      1.17.1               hfafb76e_3    conda-forge
     lcms2                     2.11                 hbd6801e_0    conda-forge
     ld_impl_linux-64          2.34                 hc38a660_9    conda-forge
     libblas                   3.8.0               17_openblas    conda-forge
     libcblas                  3.8.0               17_openblas    conda-forge
     libcurl                   7.71.1               hcdd3856_5    conda-forge
     libedit                   3.1.20191231         he28a2e2_2    conda-forge
     libev                     4.33                 h516909a_1    conda-forge
     libevent                  2.1.10               hcdb4288_2    conda-forge
     libffi                    3.2.1             he1b5a44_1007    conda-forge
     libgcc-ng                 9.3.0               h24d8f2e_16    conda-forge
     libgfortran-ng            7.5.0               hdf63c60_16    conda-forge
     libgomp                   9.3.0               h24d8f2e_16    conda-forge
     liblapack                 3.8.0               17_openblas    conda-forge
     libllvm10                 10.0.1               he513fc3_3    conda-forge
     libllvm8                  8.0.1                hc9558a2_0    conda-forge
     libnghttp2                1.41.0               h8cfc5f6_2    conda-forge
     libopenblas               0.3.10          pthreads_hb3c22a3_4    conda-forge
     libpng                    1.6.37               hed695b0_2    conda-forge
     libprotobuf               3.12.4               h8b12597_0    conda-forge
     librmm                    0.16.0a200909   cuda10.2_g9d02c5b_365    rapidsai-nightly
     libsodium                 1.0.18               h516909a_0    conda-forge
     libssh2                   1.9.0                hab1572f_5    conda-forge
     libstdcxx-ng              9.3.0               hdf63c60_16    conda-forge
     libtiff                   4.1.0                hc7e4089_6    conda-forge
     libutf8proc               2.5.0                h516909a_2    conda-forge
     libuv                     1.39.0               h516909a_0    conda-forge
     libwebp-base              1.1.0                h516909a_3    conda-forge
     llvmlite                  0.34.0           py37h5202443_1    conda-forge
     locket                    0.2.0                      py_2    conda-forge
     lz4-c                     1.9.2                he1b5a44_3    conda-forge
     markdown                  3.2.2                      py_0    conda-forge
     markupsafe                1.1.1            py37h8f50634_1    conda-forge
     mccabe                    0.6.1                      py_1    conda-forge
     mimesis                   4.0.0              pyh9f0ad1d_0    conda-forge
     mistune                   0.8.4           py37h8f50634_1001    conda-forge
     more-itertools            8.5.0                      py_0    conda-forge
     msgpack-python            1.0.0            py37h99015e2_1    conda-forge
     nbconvert                 5.6.1            py37hc8dfbb8_1    conda-forge
     nbformat                  5.0.7                      py_0    conda-forge
     nbsphinx                  0.7.1              pyh9f0ad1d_0    conda-forge
     nccl                      2.7.8.1              hc6a2c23_0    conda-forge
     ncurses                   6.2                  he1b5a44_1    conda-forge
     nodeenv                   1.5.0              pyh9f0ad1d_0    conda-forge
     notebook                  6.1.3            py37hc8dfbb8_0    conda-forge
     numba                     0.51.2           py37h9fdb41a_0    conda-forge
     numpy                     1.19.1           py37h7ea13bd_2    conda-forge
     numpydoc                  1.1.0              pyh9f0ad1d_0    conda-forge
     olefile                   0.46                       py_0    conda-forge
     openssl                   1.1.1g               h516909a_1    conda-forge
     packaging                 20.4               pyh9f0ad1d_0    conda-forge
     pandas                    1.1.2            py37h3340039_0    conda-forge
     pandoc                    1.19.2                        0    conda-forge
     pandocfilters             1.4.2                      py_1    conda-forge
     parquet-cpp               1.5.1                         2    conda-forge
     parso                     0.7.1              pyh9f0ad1d_0    conda-forge
     partd                     1.1.0                      py_0    conda-forge
     pathspec                  0.8.0              pyh9f0ad1d_0    conda-forge
     pexpect                   4.8.0            py37hc8dfbb8_1    conda-forge
     pickleshare               0.7.5           py37hc8dfbb8_1001    conda-forge
     pillow                    7.2.0            py37h718be6c_1    conda-forge
     pip                       20.2.3                     py_0    conda-forge
     pluggy                    0.13.1           py37hc8dfbb8_2    conda-forge
     pre-commit                2.7.1            py37hc8dfbb8_0    conda-forge
     pre_commit                2.7.1                         0    conda-forge
     prometheus_client         0.8.0              pyh9f0ad1d_0    conda-forge
     prompt-toolkit            3.0.7                      py_0    conda-forge
     psutil                    5.7.2            py37h8f50634_0    conda-forge
     ptyprocess                0.6.0                   py_1001    conda-forge
     py                        1.9.0              pyh9f0ad1d_0    conda-forge
     pyarrow                   1.0.1           py37h1234567_1_cuda    conda-forge
     pycodestyle               2.6.0              pyh9f0ad1d_0    conda-forge
     pycparser                 2.20               pyh9f0ad1d_2    conda-forge
     pyflakes                  2.2.0              pyh9f0ad1d_0    conda-forge
     pygments                  2.6.1                      py_0    conda-forge
     pyopenssl                 19.1.0                     py_1    conda-forge
     pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
     pyrsistent                0.16.0           py37h8f50634_0    conda-forge
     pysocks                   1.7.1            py37hc8dfbb8_1    conda-forge
     pytest                    6.0.1            py37hc8dfbb8_0    conda-forge
     python                    3.7.8           h6f2ec95_1_cpython    conda-forge
     python-dateutil           2.8.1                      py_0    conda-forge
     python_abi                3.7                     1_cp37m    conda-forge
     pytz                      2020.1             pyh9f0ad1d_0    conda-forge
     pyyaml                    5.3.1            py37h8f50634_0    conda-forge
     pyzmq                     19.0.2           py37hac76be4_0    conda-forge
     rapidjson                 1.1.0             he1b5a44_1002    conda-forge
     re2                       2020.08.01           he1b5a44_0    conda-forge
     readline                  8.0                  he28a2e2_2    conda-forge
     recommonmark              0.6.0                      py_0    conda-forge
     regex                     2020.7.14        py37h8f50634_0    conda-forge
     requests                  2.24.0             pyh9f0ad1d_0    conda-forge
     rhash                     1.3.6             h14c3975_1001    conda-forge
     rmm                       0.16.0a200909   cuda_10.2_py37_g9d02c5b_365    rapidsai-nightly
     send2trash                1.5.0                      py_0    conda-forge
     setuptools                49.6.0           py37hc8dfbb8_0    conda-forge
     six                       1.15.0             pyh9f0ad1d_0    conda-forge
     snappy                    1.1.8                he1b5a44_3    conda-forge
     snowballstemmer           2.0.0                      py_0    conda-forge
     sortedcontainers          2.2.2              pyh9f0ad1d_0    conda-forge
     spdlog                    1.7.0                hc9558a2_2    conda-forge
     sphinx                    3.2.1                      py_0    conda-forge
     sphinx-copybutton         0.3.0              pyh9f0ad1d_0    conda-forge
     sphinx-markdown-tables    0.0.14             pyh9f0ad1d_1    conda-forge
     sphinx_rtd_theme          0.5.0              pyh9f0ad1d_0    conda-forge
     sphinxcontrib-applehelp   1.0.2                      py_0    conda-forge
     sphinxcontrib-devhelp     1.0.2                      py_0    conda-forge
     sphinxcontrib-htmlhelp    1.0.3                      py_0    conda-forge
     sphinxcontrib-jsmath      1.0.1                      py_0    conda-forge
     sphinxcontrib-qthelp      1.0.3                      py_0    conda-forge
     sphinxcontrib-serializinghtml 1.1.4                      py_0    conda-forge
     sphinxcontrib-websupport  1.2.4              pyh9f0ad1d_0    conda-forge
     sqlite                    3.33.0               h4cf870e_0    conda-forge
     streamz                   0.5.5                    pypi_0    pypi
     tblib                     1.6.0                      py_0    conda-forge
     terminado                 0.8.3            py37hc8dfbb8_1    conda-forge
     testpath                  0.4.4                      py_0    conda-forge
     thrift-cpp                0.13.0               h62aa4f2_3    conda-forge
     tk                        8.6.10               hed695b0_0    conda-forge
     toml                      0.10.1             pyh9f0ad1d_0    conda-forge
     toolz                     0.10.0                     py_0    conda-forge
     tornado                   6.0.4            py37h8f50634_1    conda-forge
     traitlets                 5.0.4                      py_0    conda-forge
     typed-ast                 1.4.1            py37h516909a_0    conda-forge
     typing_extensions         3.7.4.2                    py_0    conda-forge
     urllib3                   1.25.10                    py_0    conda-forge
     virtualenv                20.0.20          py37hc8dfbb8_1    conda-forge
     wcwidth                   0.2.5              pyh9f0ad1d_1    conda-forge
     webencodings              0.5.1                      py_1    conda-forge
     wheel                     0.35.1             pyh9f0ad1d_0    conda-forge
     xz                        5.2.5                h516909a_1    conda-forge
     yaml                      0.2.5                h516909a_0    conda-forge
     zeromq                    4.3.2                he1b5a44_3    conda-forge
     zict                      2.0.0                      py_0    conda-forge
     zipp                      3.1.0                      py_0    conda-forge
     zlib                      1.2.11            h516909a_1009    conda-forge
     zstd                      1.4.5                h6597ccf_2    conda-forge
     
</pre></details>

### PR DESCRIPTION
**Pull request by davidwendt**
Closes #6219 

Strings columns that contain varying fractional seconds digits may not parse correctly since the number of digits is defined by the `%f` as documented here: https://docs.rapids.ai/api/libcudf/stable/group__strings__convert.html#gacaf50d3b6d14ece094b9fc16b3d35c76
However, when the `%f` specifier appears at the end of the format string, then it is possible to infer the remaining digits of a truncated string are `0`
For example, the following two strings specify the same time: `"03.010000", "03.01"` in seconds though the 2nd string is invalid since the `%f` specifier requires 6 numeric digits (characters).

This PR modifies the handling of `%f` to calculate correctly if the end of the string is reached before the number of digits runs out.
Also added a gtest for this case in this PR.

### CODE REVIEW

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

File: cpp/src/strings/convert/convert_datetime.cu [hunk 1]
```diff
@@ -178,6 +179,18 @@ struct parse_datetime {
...
+   * @brief Return power of ten value given an exponent.
+   *
+   * @return `1x10^exponent` for `0 <= exponent <= 9`
+   */
+  __device__ int64_t power_of_ten(int32_t exponent)
+  {
+    static int64_t powers_of_ten[] = {
+      1L, 10L, 100L, 1000L, 10000L, 100000L, 1000000L, 10000000L, 100000000L, 1000000000L};
+    return powers_of_ten[exponent];
+  }
```
Comment (codereport) [inline cpp/src/strings/convert/convert_datetime.cu:192]:
```suggestion
  __device__ constexpr int64_t power_of_ten(int32_t exponent)
  {
    constexpr int64_t powers_of_ten[] = {
      1L, 10L, 100L, 1000L, 10000L, 100000L, 1000000L, 10000000L, 100000000L, 1000000000L};
    return powers_of_ten[exponent];
  }
```

File: cpp/src/strings/convert/convert_datetime.cu [hunk 2]
```diff
@@ -198,8 +211,9 @@ struct parse_datetime {
     auto ptr    = d_string.data();
     auto length = d_string.size_bytes();
```
Comment (codereport) [inline cpp/src/strings/convert/convert_datetime.cu:212]:
```suggestion
    auto const length = d_string.size_bytes();
```

File: cpp/src/strings/convert/convert_datetime.cu [hunk 3]
```diff
@@ -219,7 +233,12 @@ struct parse_datetime {
         case 'I': timeparts[TP_HOUR] = str2int(ptr, item.length); break;
         case 'M': timeparts[TP_MINUTE] = str2int(ptr, item.length); break;
         case 'S': timeparts[TP_SECOND] = str2int(ptr, item.length); break;
-        case 'f': timeparts[TP_SUBSECOND] = str2int(ptr, item.length); break;
+        case 'f': {
+          int32_t read_size       = std::min(static_cast<int32_t>(item.length), length);
+          int64_t fraction        = str2int(ptr, read_size) * power_of_ten(item.length - read_size);
```
Comment (codereport) [inline cpp/src/strings/convert/convert_datetime.cu:238]:
```suggestion
          int32_t const read_size       = std::min(static_cast<int32_t>(item.length), length);
          int64_t const fraction        = str2int(ptr, read_size) * power_of_ten(item.length - read_size);
```

**Review by codereport — CHANGES_REQUESTED** [review]

**Review by karthikeyann — APPROVED** [review]

File: cpp/src/strings/convert/convert_datetime.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/strings/convert/convert_datetime.cu:212]: This is a working variable that is modified inside the for-loop just that starts just below this line.

**Review by codereport — APPROVED** [review]
Looks good. I would maybe add a couple more unit test values with subsecond values other than `..010..`, maybe some values like `0098700` and `0023456`

**Pull request merged**